File: ios/Classes/ExtendedPlatformView.swift
//
//  ExtendedPlatformView.swift
//  extended_platform_view
//
//  Created by 砂賀開晴 on 2022/08/27.
//

import Foundation
import Flutter

public typealias ExtendedPlatformViewBuilder = () -> ExtendedPlatformView

public struct CreationParams {
    public let frame: CGRect
    public let viewId: Int64
    public let args: Any?
}

open class ExtendedPlatformView {
    private (set) public var binaryMessenger: FlutterBinaryMessenger!
    private (set) public var methodChannel: FlutterMethodChannel!
    private (set) public var view: UIView!
    
    public init() {}
    
    func initInternal(messenger: FlutterBinaryMessenger, viewType: String, id: Int64, params: CreationParams) {
        binaryMessenger = messenger
        methodChannel = FlutterMethodChannel(name: "\(viewType)/\(id)/method", binaryMessenger: messenger)
        view = initialize(params: params)
    }
    
    open func initialize(params: CreationParams) -> UIView {
        fatalError("initialize is not implmented")
    }
}

class ExtendedPlatformViewFactoryWrapper: NSObject, FlutterPlatformViewFactory {
    init(messenger: FlutterBinaryMessenger, viewType: String, builder: @escaping ExtendedPlatformViewBuilder) {
        self.messenger = messenger
        self.viewType = viewType
        self.builder = builder
        super.init()
    }
    
    private let messenger: FlutterBinaryMessenger
    private let viewType: String
    private let builder: ExtendedPlatformViewBuilder
    
    func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
        let params = CreationParams(frame: frame, viewId: viewId, args: args)
        let view = builder()
        view.initInternal(messenger: messenger, viewType: viewType, id: viewId, params: params)
        return ExtendedPlatformViewWrapper(view: view)
    }
    
    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }
}

class ExtendedPlatformViewWrapper: NSObject, FlutterPlatformView {
    init(view: ExtendedPlatformView) {
        platformView = view
        super.init()
    }
    
    private let platformView: ExtendedPlatformView
    
    func view() -> UIView {
        return platformView.view
    }
}

File: android/src/main/kotlin/work/kscafe/extended_platform_view/ExtendedPlatformView.kt
package work.kscafe.extended_platform_view

import android.annotation.SuppressLint
import android.content.Context
import android.view.View
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

typealias ExtendedPlatformViewBuilder = () -> ExtendedPlatformView

data class CreationParams(val viewId: Int, val args: Any?)

abstract class ExtendedPlatformView() {
    lateinit var context: Context
        private set
    lateinit var binaryMessenger: BinaryMessenger
        private set
    lateinit var methodChannel: MethodChannel
        private set
    lateinit var view: View
        private set

    internal fun initInternal(context: Context, messenger: BinaryMessenger, viewType: String, id: Int, params: CreationParams) {
        this.context = context
        binaryMessenger = messenger
        methodChannel = MethodChannel(messenger, "${viewType}/${id}/method")
        view = initialize(params)
    }

    abstract fun initialize(params: CreationParams): View

    fun onFlutterViewAttached(flutterView: View) {}
    fun onFlutterViewDetached() {}
    fun dispose() {}
}

internal class ExtendedPlatformViewFactoryWrapper(
    private val messenger: BinaryMessenger,
    private val viewType: String,
    private val builder: ExtendedPlatformViewBuilder): PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    override fun create(context: Context?, viewId: Int, args: Any?): PlatformView {
        val view = builder().also {
            val params = CreationParams(viewId, args)

            // MEMO: force unwrapping context. see: https://github.com/flutter/flutter/issues/104480
            it.initInternal(context!!, messenger, viewType, viewId, params)
        }
        return ExtendedPlatformViewWrapper(view)
    }

    internal class ExtendedPlatformViewWrapper(private val platformView: ExtendedPlatformView): PlatformView {
        override fun getView(): View {
            return platformView.view
        }

        override fun dispose() {
            platformView.dispose()
        }

        override fun onFlutterViewAttached(flutterView: View) {
            platformView.onFlutterViewAttached(flutterView)
        }

        override fun onFlutterViewDetached() {
            platformView.onFlutterViewDetached()
        }
    }
}